File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun isValidBST(root: TreeNode?): Boolean {
        return validate(root, Long.MIN_VALUE, Long.MAX_VALUE)
    }

    private fun validate(node: TreeNode?, left: Long, right: Long): Boolean {
        node ?: return true

        val value = node.`val`.toLong()
        if (value <= left || value >= right) {
            return false
        }

        return validate(node.left, left, value) && validate(node.right, value, right)
    }
}

File: Data Structures & Algorithms/invert-a-binary-tree/submission-5.py
# BFS
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return None

        queue = deque([root])
        while queue:
            node = queue.popleft()
            temp = node.left
            node.left = node.right
            node.right = temp
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)
        
        return root
